packages/bot-web-ui/src/stores/transactions-store.js

```diff
@@ -172,37 +173,45 @@ export default class TransactionsStore {
         });
     }
 
-    recoverPendingContractsById(contract_id) {
-        const { ws } = this.root_store;
-
-        ws.authorized.subscribeProposalOpenContract(contract_id, response => {
-            if (!response.error) {
-                const { proposal_open_contract } = response;
-
-                const { contract_info } = this.root_store.summary_card;
+    updateResultsCompletedContract(contract) {
+        const { journal, summary_card } = this.root_store;
+        const { contract_info } = summary_card;
+        const { currency, profit } = contract;
 
-                if (proposal_open_contract.contract_id === contract_info?.contract_id) return;
+        if (contract.contract_id !== contract_info?.contract_id) {
+            this.onBotContractEvent(contract);
 
-                this.onBotContractEvent(proposal_open_contract);
-
-                if (!this.recovered_transactions.includes(proposal_open_contract.contract_id)) {
-                    this.recovered_transactions.push(proposal_open_contract.contract_id);
-                }
+            if (!this.recovered_transactions.includes(contract.contract_id)) {
+                this.recovered_transactions.push(contract.contract_id);
+            }
+            if (!this.recovered_completed_transactions.includes(contract.contract_id) && isEnded(contract)) {
+                this.recovered_completed_transactions.push(contract.contract_id);
 
-                if (
-                    !this.recovered_completed_transactions.includes(proposal_open_contract.contract_id) &&
-                    isEnded(proposal_open_contract)
-                ) {
-                    this.recovered_completed_transactions.push(proposal_open_contract.contract_id);
+                journal.onLogSuccess({
+                    log_type: profit > 0 ? log_types.PROFIT : log_types.LOST,
+                    extra: { currency, profit },
+                });
+            }
+        }
+    }
 
-                    const { currency, profit } = proposal_open_contract;
+    recoverPendingContractsById(contract_id) {
+        const { ws, core } = this.root_store;
+        const positions = core.portfolio.positions;
 
-                    this.root_store.journal.onLogSuccess({
-                        log_type: profit > 0 ? log_types.PROFIT : log_types.LOST,
-                        extra: { currency, profit },
-                    });
-                }
+        ws.authorized.subscribeProposalOpenContract(contract_id, response => {
+            this.is_called_proposal_open_contract = true;
+            if (!response.error) {
+                const { proposal_open_contract } = response;
+                this.updateResultsCompletedContract(proposal_open_contract);
             }
         });
+
+        if (!this.is_called_proposal_open_contract) {
```

Comment:
I think this line of code is executed before deciding to change the  is_called_proposal_open_contract to true in the subscribeProposalOpenContract. Is it expected?

Reply:
No, this line is executed only when the `is_called_proposal_open_contract = true`. You can check it with console.log inside of this block of code and with console.log inside of this block **ws.authorized.subscribeProposalOpenContract(contract_id, response => {**. Everything works as expected.

Reply:
Lemme explain my meaning
we have :
    is_called_proposal_open_contract = false;

    ws.authorized.subscribeProposalOpenContract(contract_id, response => {
            this.is_called_proposal_open_contract = true;
            if (!response.error) {
                const { proposal_open_contract } = response;
                this.updateResultsCompletedContract(proposal_open_contract);
            }
        });

 if (!this.is_called_proposal_open_contract) {

     ...
  }
So this line:
`this.is_called_proposal_open_contract = true;`
only happens when the response is ready … during waiting for the response, the next line is being executed and because it is initialized with false so it goes through the if  statement.
because it is asynchronous.

Reply:
@sara-fs  no async problem, if we still wait for the response for a current contract with an id of contract then it means that contract wasn't completed. if it wasn't completed, then on the bot it will show loading. Because with the data from the store we check there in the method is completed or not